Felix fails to program rules for ICMP port

Looks like a mismatch between plugin and felix: the plugin is specifying both icmp_type and ports.  Also looks like felix needs to handle multiple ICMP ports and port ranges, which we may not at present (just spotted that multiport doesn't support icmp, d'oh).

```
Apr 18 19:58:33.553531 10.41.0.18 err calico-felix: Failed to convert rules to fragments: [{'ip_version': 4, 'src_tag': u'ed1c1daf-8cb8-44c8-a2cc-6ca518f31301', 'icmp_type': 255, 'protocol': u'icmp', 'dst_ports': [255]}].  Will DROP!
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/calico/felix/frules.py", line 153, in rules_to_chain_rewrite_lines
  File "/usr/lib64/python2.7/site-packages/calico/felix/frules.py", line 209, in rule_to_iptables_fragments
  File "/usr/lib64/python2.7/site-packages/calico/felix/frules.py", line 295, in _rule_to_iptables_fragment
AssertionError: Protocol icmp not supported with dst_ports ({'protocol': u'icmp', 'dst_ports': ['255'], 'src_tag': u'ed1c1daf-8cb8-44c8-a2cc-6ca518f31301', 'ip_version': 4, 'icmp_type': 255, 'src_ports': []})
```

--- a/calico/felix/fetcd.py
+++ b/calico/felix/fetcd.py
@@ -517,7 +517,17 @@ def validate_rules(rules):
                 issues.append("Invalid action in rule %s." % rule)
 
             icmp_type = rule.get('icmp_type')
-            #TODO: firewall the icmp_type too
+            if icmp_type is not None:
+                if not 0 <= icmp_type <= 255:
+                    issues.append("ICMP type is out of range.")
+            icmp_code = rule.get("icmp_code")
+            if icmp_code is not None:
+                if not 0 <= icmp_code <= 255:
+                    issues.append("ICMP code is out of range.")
+                if icmp_type is None:
+                    # FIXME ICMP code without ICMP type not supported by iptables
+                    # Firewall against that for now.
+                    issues.append("ICMP code specified without ICMP type.")
 
     if issues:
         raise ValidationFailed(" ".join(issues))

--- a/calico/felix/frules.py
+++ b/calico/felix/frules.py
@@ -299,15 +299,21 @@ def _rule_to_iptables_fragment(chain_name, rule, ip_version, tag_to_ipset,
             assert ports.count(",") + ports.count(":") < 15, "Too many ports"
             append("--match multiport", "--%s-ports" % direction, ports)
 
-    if "icmp_type" in rule and rule["icmp_type"] is not None:
+    if rule.get("icmp_type") is not None:
         icmp_type = rule["icmp_type"]
         assert isinstance(icmp_type, int), "ICMP type should be an int"
+        if "icmp_code" in rule:
+            icmp_code = rule["icmp_code"]
+            assert isinstance(icmp_code, int), "ICMP code should be an int"
+            icmp_filter = "%s/%s" % (icmp_type, icmp_code)
+        else:
+            icmp_filter = icmp_type
         if proto == "icmp" and ip_version == 4:
-            append("--match icmp", "--icmp-type", rule["icmp_type"])
+            append("--match icmp", "--icmp-type", icmp_filter)
         elif ip_version == 6:
             assert proto == "icmpv6"
             # Note variant spelling of icmp[v]6
-            append("--match icmp6", "--icmpv6-type", rule["icmp_type"])
+            append("--match icmp6", "--icmpv6-type", icmp_filter)
 
     # Add the action
     append("--jump", on_allow if rule.get("action", "allow") == "allow"

--- a/calico/openstack/t_etcd.py
+++ b/calico/openstack/t_etcd.py
@@ -270,75 +270,11 @@ class CalicoTransportEtcd(CalicoTransport):
         for sgid in self.profile_tags(profile_id):
             for rule in self.sgs[sgid]['security_group_rules']:
                 LOG.info("Neutron rule  %s : %s", profile_id, rule)
-
-                ethertype = rule['ethertype']
-                etcd_rule = {}
-
-                # Map the ethertype field from Neutron to etcd format.
-                etcd_rule['ip_version'] = {'IPv4': 4,
-                                           'IPv6': 6}[ethertype]
-
-                # Map the protocol field from Neutron to etcd format.
-                if rule['protocol'] is None or rule['protocol'] == -1:
-                    pass
-                elif rule['protocol'] == 'icmp':
-                    etcd_rule['protocol'] = {'IPv4': 'icmp',
-                                             'IPv6': 'icmpv6'}[ethertype]
-                else:
-                    etcd_rule['protocol'] = rule['protocol']
-
-                # OpenStack (sometimes) represents 'any IP address' by setting
-                # both 'remote_group_id' and 'remote_ip_prefix' to None.  We
-                # translate that to an explicit 0.0.0.0/0 (for IPv4) or ::/0
-                # (for IPv6).
-                net = rule['remote_ip_prefix']
-                if not (net or rule['remote_group_id']):
-                    net = {'IPv4': '0.0.0.0/0',
-                           'IPv6': '::/0'}[ethertype]
-
-                port_spec = None
-                if rule['protocol'] == 'icmp':
-                    # OpenStack stashes the ICMP match criteria in
-                    # port_range_min/max.
-                    icmp_type = rule['port_range_min']
-                    if icmp_type is not None and icmp_type != -1:
-                        etcd_rule['icmp_type'] = icmp_type
-                    icmp_code = rule['port_range_max']
-                    if icmp_code is not None and icmp_code != -1:
-                        etcd_rule['icmp_code'] = icmp_code
-                else:
-
-                    # src/dst_ports is a list in which each entry can be a
-                    # single number, or a string describing a port range.
-                    if rule['port_range_min'] == -1:
-                        port_spec = ['1:65535']
-                    elif rule['port_range_min'] == rule['port_range_max']:
-                        if rule['port_range_min'] is not None:
-                            port_spec = [rule['port_range_min']]
-                    else:
-                        port_spec = ['%s:%s' % (rule['port_range_min'],
-                                                rule['port_range_max'])]
-
-                # Put it all together and add to either the inbound or the
-                # outbound list.
+                etcd_rule = _neutron_rule_to_etcd_rule(rule)
                 if rule['direction'] == 'ingress':
-                    if rule['remote_group_id'] is not None:
-                        etcd_rule['src_tag'] = rule['remote_group_id']
-                    if net is not None:
-                        etcd_rule['src_net'] = net
-                    if port_spec is not None:
-                        etcd_rule['dst_ports'] = port_spec
                     inbound.append(etcd_rule)
-                    LOG.info("=> Inbound Calico rule %s" % etcd_rule)
                 else:
-                    if rule['remote_group_id'] is not None:
-                        etcd_rule['dst_tag'] = rule['remote_group_id']
-                    if net is not None:
-                        etcd_rule['dst_net'] = net
-                    if port_spec is not None:
-                        etcd_rule['dst_ports'] = port_spec
                     outbound.append(etcd_rule)
-                    LOG.info("=> Outbound Calico rule %s" % etcd_rule)
 
         return {'inbound_rules': inbound, 'outbound_rules': outbound}
 
@@ -407,3 +343,73 @@ class CalicoTransportEtcd(CalicoTransport):
             # TODO Set this flag only once we're really ready!
             LOG.info('%s -> true', READY_KEY)
             self.client.write(READY_KEY, 'true')
+
+def _neutron_rule_to_etcd_rule(rule):
+    """
+    Translate a single Neutron rule dict to a single dict in our
+    etcd format.
+    """
+    ethertype = rule['ethertype']
+    etcd_rule = {}
+    # Map the ethertype field from Neutron to etcd format.
+    etcd_rule['ip_version'] = {'IPv4': 4,
+                               'IPv6': 6}[ethertype]
+    # Map the protocol field from Neutron to etcd format.
+    if rule['protocol'] is None or rule['protocol'] == -1:
+        pass
+    elif rule['protocol'] == 'icmp':
+        etcd_rule['protocol'] = {'IPv4': 'icmp',
+                                 'IPv6': 'icmpv6'}[ethertype]
+    else:
+        etcd_rule['protocol'] = rule['protocol']
+
+    # OpenStack (sometimes) represents 'any IP address' by setting
+    # both 'remote_group_id' and 'remote_ip_prefix' to None.  We
+    # translate that to an explicit 0.0.0.0/0 (for IPv4) or ::/0
+    # (for IPv6).
+    net = rule['remote_ip_prefix']
+    if not (net or rule['remote_group_id']):
+        net = {'IPv4': '0.0.0.0/0',
+               'IPv6': '::/0'}[ethertype]
+    port_spec = None
+    if rule['protocol'] == 'icmp':
+        # OpenStack stashes the ICMP match criteria in
+        # port_range_min/max.
+        icmp_type = rule['port_range_min']
+        if icmp_type is not None and icmp_type != -1:
+            etcd_rule['icmp_type'] = icmp_type
+        icmp_code = rule['port_range_max']
+        if icmp_code is not None and icmp_code != -1:
+            etcd_rule['icmp_code'] = icmp_code
+    else:
+        # src/dst_ports is a list in which each entry can be a
+        # single number, or a string describing a port range.
+        if rule['port_range_min'] == -1:
+            port_spec = ['1:65535']
+        elif rule['port_range_min'] == rule['port_range_max']:
+            if rule['port_range_min'] is not None:
+                port_spec = [rule['port_range_min']]
+        else:
+            port_spec = ['%s:%s' % (rule['port_range_min'],
+                                    rule['port_range_max'])]
+
+    # Put it all together and add to either the inbound or the
+    # outbound list.
+    if rule['direction'] == 'ingress':
+        if rule['remote_group_id'] is not None:
+            etcd_rule['src_tag'] = rule['remote_group_id']
+        if net is not None:
+            etcd_rule['src_net'] = net
+        if port_spec is not None:
+            etcd_rule['dst_ports'] = port_spec
+        LOG.info("=> Inbound Calico rule %s" % etcd_rule)
+    else:
+        if rule['remote_group_id'] is not None:
+            etcd_rule['dst_tag'] = rule['remote_group_id']
+        if net is not None:
+            etcd_rule['dst_net'] = net
+        if port_spec is not None:
+            etcd_rule['dst_ports'] = port_spec
+        LOG.info("=> Outbound Calico rule %s" % etcd_rule)
+
+    return etcd_rule

--- a/calico/openstack/t_etcd.py
+++ b/calico/openstack/t_etcd.py
@@ -284,7 +284,6 @@ class CalicoTransportEtcd(CalicoTransport):
                 elif rule['protocol'] == 'icmp':
                     etcd_rule['protocol'] = {'IPv4': 'icmp',
                                              'IPv6': 'icmpv6'}[ethertype]
-                    etcd_rule['icmp_type'] = rule['port_range_min']
                 else:
                     etcd_rule['protocol'] = rule['protocol']
 
@@ -297,18 +296,28 @@ class CalicoTransportEtcd(CalicoTransport):
                     net = {'IPv4': '0.0.0.0/0',
                            'IPv6': '::/0'}[ethertype]
 
-                # src/dst_ports is a list in which each entry can be a single
-                # number, or a string describing a port range.
-                if rule['port_range_min'] == -1:
-                    port_spec = ['1:65535']
-                elif rule['port_range_min'] == rule['port_range_max']:
-                    if rule['port_range_min'] is not None:
-                        port_spec = [rule['port_range_min']]
-                    else:
-                        port_spec = None
+                port_spec = None
+                if rule['protocol'] == 'icmp':
+                    # OpenStack stashes the ICMP match criteria in
+                    # port_range_min/max.
+                    icmp_type = rule['port_range_min']
+                    if icmp_type is not None and icmp_type != -1:
+                        etcd_rule['icmp_type'] = icmp_type
+                    icmp_code = rule['port_range_max']
+                    if icmp_code is not None and icmp_code != -1:
+                        etcd_rule['icmp_code'] = icmp_code
                 else:
-                    port_spec = ['%s:%s' % (rule['port_range_min'],
-                                            rule['port_range_max'])]
+
+                    # src/dst_ports is a list in which each entry can be a
+                    # single number, or a string describing a port range.
+                    if rule['port_range_min'] == -1:
+                        port_spec = ['1:65535']
+                    elif rule['port_range_min'] == rule['port_range_max']:
+                        if rule['port_range_min'] is not None:
+                            port_spec = [rule['port_range_min']]
+                    else:
+                        port_spec = ['%s:%s' % (rule['port_range_min'],
+                                                rule['port_range_max'])]
 
                 # Put it all together and add to either the inbound or the
                 # outbound list.

--- a/calico/openstack/test/test_plugin_etcd.py
+++ b/calico/openstack/test/test_plugin_etcd.py
@@ -22,6 +22,7 @@ import eventlet
 import json
 import mock
 import unittest
+from calico.felix import fetcd
 
 import calico.openstack.test.lib as lib
 import calico.openstack.mech_calico as mech_calico
@@ -547,6 +548,15 @@ class TestPluginEtcd(lib.Lib, unittest.TestCase):
     def assertNeutronToEtcd(self, neutron_rule, exp_etcd_rule):
         etcd_rule = t_etcd._neutron_rule_to_etcd_rule(neutron_rule)
         self.assertEqual(etcd_rule, exp_etcd_rule)
+
+        # Check felix is happy with generated rule.
+        if neutron_rule["direction"] == "ingress":
+            rules = {"inbound_rules": [neutron_rule],
+                     "outbound_rules": []}
+        else:
+            rules = {"outbound_rules": [neutron_rule],
+                     "inbound_rules": []}
+        fetcd.validate_rules(rules)
 
 
 def neutron_rule(overrides):

--- a/calico/openstack/test/test_plugin_etcd.py
+++ b/calico/openstack/test/test_plugin_etcd.py
@@ -495,3 +495,69 @@ class TestPluginEtcd(lib.Lib, unittest.TestCase):
             {mech_calico.api.NETWORK_TYPE: 'vlan'},
             mech_calico.constants.AGENT_TYPE_DHCP
         ))
+
+    def test_neutron_rule_to_etcd_rule_icmp(self):
+        # No type/code specified
+        self.assertNeutronToEtcd(neutron_rule({
+            "ethertype": "IPv4",
+            "protocol": "icmp",
+        }), {
+            'ip_version': 4,
+            'protocol': 'icmp',
+            'src_net': '0.0.0.0/0',
+        })
+        # Type/code wildcarded, same as above.
+        self.assertNeutronToEtcd(neutron_rule({
+            "ethertype": "IPv4",
+            "protocol": "icmp",
+            "port_range_min": -1,
+            "port_range_max": -1,
+        }), {
+            'ip_version': 4,
+            'protocol': 'icmp',
+            'src_net': '0.0.0.0/0',
+        })
+        # Type and code.
+        self.assertNeutronToEtcd(neutron_rule({
+            "ethertype": "IPv4",
+            "protocol": "icmp",
+            "port_range_min": 123,
+            "port_range_max": 100,
+        }), {
+            'ip_version': 4,
+            'protocol': 'icmp',
+            'src_net': '0.0.0.0/0',
+            'icmp_type': 123,
+            'icmp_code': 100,
+        })
+        # Type and code, IPv6.
+        self.assertNeutronToEtcd(neutron_rule({
+            "ethertype": "IPv6",
+            "protocol": "icmp",
+            "port_range_min": 123,
+            "port_range_max": 100,
+        }), {
+            'ip_version': 6,
+            'protocol': 'icmpv6',
+            'src_net': '::/0',
+            'icmp_type': 123,
+            'icmp_code': 100,
+        })
+
+    def assertNeutronToEtcd(self, neutron_rule, exp_etcd_rule):
+        etcd_rule = t_etcd._neutron_rule_to_etcd_rule(neutron_rule)
+        self.assertEqual(etcd_rule, exp_etcd_rule)
+
+
+def neutron_rule(overrides):
+    rule = {
+        "ethertype": "IPv4",
+        "protocol": None,
+        "remote_ip_prefix": None,
+        "remote_group_id": None,
+        "direction": "ingress",
+        "port_range_min": None,
+        "port_range_max": None,
+    }
+    rule.update(overrides)
+    return rule

--- a/calico/openstack/test/test_plugin_etcd.py
+++ b/calico/openstack/test/test_plugin_etcd.py
@@ -499,7 +499,7 @@ class TestPluginEtcd(lib.Lib, unittest.TestCase):
 
     def test_neutron_rule_to_etcd_rule_icmp(self):
         # No type/code specified
-        self.assertNeutronToEtcd(neutron_rule({
+        self.assertNeutronToEtcd(_neutron_rule_from_dict({
             "ethertype": "IPv4",
             "protocol": "icmp",
         }), {
@@ -508,7 +508,7 @@ class TestPluginEtcd(lib.Lib, unittest.TestCase):
             'src_net': '0.0.0.0/0',
         })
         # Type/code wildcarded, same as above.
-        self.assertNeutronToEtcd(neutron_rule({
+        self.assertNeutronToEtcd(_neutron_rule_from_dict({
             "ethertype": "IPv4",
             "protocol": "icmp",
             "port_range_min": -1,
@@ -519,7 +519,7 @@ class TestPluginEtcd(lib.Lib, unittest.TestCase):
             'src_net': '0.0.0.0/0',
         })
         # Type and code.
-        self.assertNeutronToEtcd(neutron_rule({
+        self.assertNeutronToEtcd(_neutron_rule_from_dict({
             "ethertype": "IPv4",
             "protocol": "icmp",
             "port_range_min": 123,
@@ -532,7 +532,7 @@ class TestPluginEtcd(lib.Lib, unittest.TestCase):
             'icmp_code': 100,
         })
         # Type and code, IPv6.
-        self.assertNeutronToEtcd(neutron_rule({
+        self.assertNeutronToEtcd(_neutron_rule_from_dict({
             "ethertype": "IPv6",
             "protocol": "icmp",
             "port_range_min": 123,
@@ -559,7 +559,7 @@ class TestPluginEtcd(lib.Lib, unittest.TestCase):
         fetcd.validate_rules(rules)
 
 
-def neutron_rule(overrides):
+def _neutron_rule_from_dict(overrides):
     rule = {
         "ethertype": "IPv4",
         "protocol": None,

